python/lsst/afw/cameraGeom/utils.py

```diff
@@ -336,64 +343,116 @@ def _prepareImage(self, ccd, im, binSize, allowRotate=True):
 
-        ccdImage = im.Factory(bbox)
+        if self.callback:
+            try:
+                im = self.callback(im, ccd, imageSource=self)
+            except Exception, e:
+                if self.verbose:
+                    print("callback failed: %s" % e)
+                im = imageFactory(*bbox.getDimensions())
 
-        ampImages = []
-        for a in ccd:
-            bias = im[a.getRawHorizontalOverscanBBox()]
+        return self._prepareImage(ccd, im, binSize, allowRotate=allowRotate), ccd
+
+def rawCallback(im, ccd=None, imageSource=None,
+                correctGain=False, subtractBias=False, convertToFloat=False):
+    """!A callback function that may or may not subtract bias/correct gain/trim a raw image
+
+    @param im   An image of a chip, ready to be binned and maybe rotated.  Maybe an Exposure,
+                a MaskedImage or an Image
+    @param ccd  The Detector; if None assume that im is an exposure and extract it's Detector
+    @param imageSource  An object derived from FakeImageDataSource (typically ButlerImage)
+    @param correctGain  Correct each amplifier for its gain
+    @param subtractBias Subtract the bias from each amplifier
+    @param convertToFloat  Convert im to floating point if possible
+
+    N.b. if imageSource is derived from ButlerImage, imageSource.butler is available
+    """
+    if ccd is None:
+        ccd = cameraGeom.cast_Ccd(im.getDetector())
+    if hasattr(im, "getMaskedImage"):
+        im = im.getMaskedImage()
+    if convertToFloat and hasattr(im, "convertF"):
+        im = im.convertF()
+
+    isTrimmed = imageSource.isTrimmed
+    if isTrimmed:
+        bbox = ccd.getBBox()
+    else:
+        bbox = calcRawCcdBBox(ccd)
+
+    ccdImage = im.Factory(bbox)
+
+    ampImages = []
+    for a in ccd:
+        if isTrimmed:
             data = im[a.getRawDataBBox()]
+        else:
+            data = im
+
+        if subtractBias:
+            bias = im[a.getRawHorizontalOverscanBBox()]
             data -= afwMath.makeStatistics(bias, afwMath.MEANCLIP).getValue()
+        if correctGain:
             data *= a.getGain()
 
-            ampImages.append(data)
+        ampImages.append(data)
 
-        ccdImage = imageFactory(bbox)
-        for ampImage, amp in zip(ampImages, ccd):
-            if self.isTrimmed:
-                assembleAmplifierImage(ccdImage, ampImage, amp)
-            else:
-                assembleAmplifierRawImage(ccdImage, ampImage, amp)
+    ccdImage = type(ccdImage)(bbox)
+    for ampImage, amp in zip(ampImages, ccd):
+        if isTrimmed:
+            assembleAmplifierImage(ccdImage, ampImage, amp)
+        else:
+            assembleAmplifierRawImage(ccdImage, ampImage, amp)
 
-        return ccdImage
+    return ccdImage
 
-def overlayCcdBoxes(ccd, untrimmedCcdBbox, nQuarter, isTrimmed, ccdOrigin, display, binSize):
+def overlayCcdBoxes(ccd, untrimmedCcdBbox=None, nQuarter=0,
```

Comment:
Should add a description of what `nQuarter` is/does, as I don't feel that the parameter name is sufficiently self-explanatory.